### PR TITLE
CIRC-1479 TLR not respecting request policies

### PR DESCRIPTION
**Pull request by bisecomsergiy**
### Purpose
Change requests sequence creation for mod-patron
### Approach

- Changed requests types sequence
- Update TLR policy condition
- Tests fixed

### Learning
https://issues.folio.org/browse/CIRC-1479

### CODE REVIEW

File: src/main/java/org/folio/circulation/domain/CreateRequestService.java [hunk 1]
```diff
@@ -140,7 +140,8 @@ private CompletableFuture<Result<RequestAndRelatedRecords>> checkPolicy(
     boolean tlrFeatureEnabled = records.getRequest().getTlrSettingsConfiguration()
       .isTitleLevelRequestsFeatureEnabled();
 
-    if (tlrFeatureEnabled && records.getRequest().getRequestLevel() == TITLE) {
+    if (tlrFeatureEnabled && records.getRequest().getRequestLevel() == TITLE
+      && records.getRequest().isHold()) {
       return completedFuture(succeeded(records));
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulation/domain/CreateRequestService.java:145]: Blank line

**Review by alexanderkurash — APPROVED** [review]

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java [hunk 2]
```diff
@@ -91,7 +91,7 @@
 
 public class RequestByInstanceIdResource extends Resource {
 
-  private static final List<RequestType> ORDERED_REQUEST_TYPES = List.of(HOLD, RECALL, PAGE);
+  private static final List<RequestType> ORDERED_REQUEST_TYPES = List.of(RECALL, HOLD, PAGE);
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java:94]: Add two test scenarios
1 - RECALL is forbidden, HOLD is allowed - HOLD should be created
2 - RECALL is forbidden, HOLD is forbidden - HOLD should be created

File: src/test/java/api/requests/InstanceRequestsAPICreationTests.java [hunk 3]
```diff
@@ -912,6 +918,60 @@ void tlrRequestCreatedWhenTlrFeatureEnabled() {
...
+    JsonObject requestBody = new RequestByInstanceIdRequestBuilder()
+      .withInstanceId(checkedOutItem.getInstanceId())
+      .withRequestDate(requestDate)
+      .withRequesterId(requesterId)
+      .withPickupServicePointId(pickupServicePointId)
+      .withRequestExpirationDate(requestDate)
+      .create();
+
+    Response requestResponse = requestsFixture.attemptToPlaceForInstance(requestBody);
+    assertEquals(201, requestResponse.getStatusCode());
```
Comment (alexanderkurash) [inline src/test/java/api/requests/InstanceRequestsAPICreationTests.java:966]: HTTP_CREATED

File: src/test/java/api/requests/InstanceRequestsAPICreationTests.java [hunk 4]
```diff
@@ -912,6 +918,60 @@ void tlrRequestCreatedWhenTlrFeatureEnabled() {
       getUserContextMatchers(requester)));
   }
 
+  @ParameterizedTest
+  @ValueSource(strings = {"HOLD,PAGE", "PAGE"})
```
Comment (alexanderkurash) [inline src/test/java/api/requests/InstanceRequestsAPICreationTests.java:922]: ""

File: src/test/java/api/requests/InstanceRequestsAPICreationTests.java [hunk 5]
```diff
@@ -912,6 +918,60 @@ void tlrRequestCreatedWhenTlrFeatureEnabled() {
...
+  void canCreateHoldTitleLevelRequestWithRequestPolicyNotAllowingRecallsOrRecallsAndHolds(String allowedRequestTypes) {
```
Comment (alexanderkurash) [inline src/test/java/api/requests/InstanceRequestsAPICreationTests.java:923]: canCreateHoldTLRWhenNotAllowedByRequestPolicy

File: src/test/java/api/requests/InstanceRequestsAPICreationTests.java [hunk 6]
```diff
@@ -912,6 +918,60 @@ void tlrRequestCreatedWhenTlrFeatureEnabled() {
...
+    UUID confirmationTemplateId = UUID.randomUUID();
+    templateFixture.createDummyNoticeTemplate(confirmationTemplateId);
+    reconfigureTlrFeature(TlrFeatureStatus.ENABLED, confirmationTemplateId, null, null);
+
+    final String undergradPatronGroupPolicy = patronGroupsFixture.undergrad().getId().toString();
+    final String anyNoticePolicy = noticePoliciesFixture.activeNotice().getId().toString();
+    final String anyLoanPolicy = loanPoliciesFixture.canCirculateRolling().getId().toString();
```
[the remark this answers is not in the document]
Comment (bisecomsergiy) [reply src/test/java/api/requests/InstanceRequestsAPICreationTests.java]: It is done, thanks

File: src/test/java/api/requests/InstanceRequestsAPICreationTests.java [hunk 7]
```diff
@@ -912,6 +918,60 @@ void tlrRequestCreatedWhenTlrFeatureEnabled() {
...
+    ArrayList<RequestType> allowedRequestPolicyTypes = new ArrayList<>();
+    Arrays.stream(allowedRequestTypes.split(",")).forEach(type -> allowedRequestPolicyTypes.add(RequestType.from(type)));
+    final String requestPolicy = requestPoliciesFixture.customRequestPolicy(allowedRequestPolicyTypes,
+      "Request policy name", "Request policy description").getId().toString();
+    final String rules = String.join("\n",
+      "priority: t, s, c, b, a, m, g",
+      "fallback-policy : l " + anyLoanPolicy + " r " + anyRequestPolicy + " n " + anyNoticePolicy + " o " + anyOverdueFinePolicy + " i " + anyLostItemFeePolicy +  "\n",
+      "g " + undergradPatronGroupPolicy + ": l " + anyLoanPolicy + " r " + requestPolicy +" n " + anyNoticePolicy + " o " + anyOverdueFinePolicy + " i " + anyLostItemFeePolicy
+    );
+    try {
```
[the remark this answers is not in the document]
Comment (bisecomsergiy) [reply src/test/java/api/requests/InstanceRequestsAPICreationTests.java]: Removed

File: src/test/java/api/requests/InstanceRequestsAPICreationTests.java [hunk 8]
```diff
@@ -912,6 +918,60 @@ void tlrRequestCreatedWhenTlrFeatureEnabled() {
...
+    final String anyOverdueFinePolicy = overdueFinePoliciesFixture.facultyStandard().getId().toString();
+    final String anyLostItemFeePolicy = lostItemFeePoliciesFixture.facultyStandard().getId().toString();
+
+    ArrayList<RequestType> allowedRequestPolicyTypes = new ArrayList<>();
+    Arrays.stream(allowedRequestTypes.split(",")).forEach(type -> allowedRequestPolicyTypes.add(RequestType.from(type)));
+    final String requestPolicy = requestPoliciesFixture.customRequestPolicy(allowedRequestPolicyTypes,
+      "Request policy name", "Request policy description").getId().toString();
+    final String rules = String.join("\n",
+      "priority: t, s, c, b, a, m, g",
+      "fallback-policy : l " + anyLoanPolicy + " r " + anyRequestPolicy + " n " + anyNoticePolicy + " o " + anyOverdueFinePolicy + " i " + anyLostItemFeePolicy +  "\n",
```
[the remark this answers is not in the document]
Comment (bisecomsergiy) [reply src/test/java/api/requests/InstanceRequestsAPICreationTests.java]: Used a method, thanks

File: src/test/java/api/requests/InstanceRequestsAPICreationTests.java (hunk 3 above)
Comment (bisecomsergiy) [reply src/test/java/api/requests/InstanceRequestsAPICreationTests.java]: Done, thanks

File: src/test/java/api/requests/InstanceRequestsAPICreationTests.java (hunk 4 above)
Comment (bisecomsergiy) [reply src/test/java/api/requests/InstanceRequestsAPICreationTests.java]: Added parameter, thanks

File: src/test/java/api/requests/InstanceRequestsAPICreationTests.java (hunk 5 above)
Comment (bisecomsergiy) [reply src/test/java/api/requests/InstanceRequestsAPICreationTests.java]: Name is changed accordingly.

File: src/test/java/api/requests/InstanceRequestsAPICreationTests.java [hunk 9]
```diff
@@ -912,6 +918,47 @@ void tlrRequestCreatedWhenTlrFeatureEnabled() {
...
+    ArrayList<RequestType> allowedRequestPolicyTypes = new ArrayList<>();
+    Arrays.stream(allowedRequestTypes.split(",")).forEach(type -> allowedRequestPolicyTypes.add(RequestType.from(type)));
+    final IndividualResource requestPolicy = requestPoliciesFixture.customRequestPolicy(allowedRequestPolicyTypes,
+      "Request policy name", "Request policy description");
+
+    useFallbackPolicies(loanPolicy.getId(),
+      requestPolicy.getId(),
+      noticePolicy.getId(),
+      overdueFinePolicy.getId(),
+      lostItemFeePolicy.getId());
```
[the remark this answers is not in the document]
Comment (bisecomsergiy) [reply src/test/java/api/requests/InstanceRequestsAPICreationTests.java]: It is done, thanks!

File: src/test/java/api/requests/InstanceRequestsAPICreationTests.java [hunk 10]
```diff
@@ -912,6 +918,47 @@ void tlrRequestCreatedWhenTlrFeatureEnabled() {
...
+    templateFixture.createDummyNoticeTemplate(confirmationTemplateId);
+    reconfigureTlrFeature(TlrFeatureStatus.ENABLED, confirmationTemplateId, null, null);
+
+    final IndividualResource noticePolicy = noticePoliciesFixture.activeNotice();
+    final IndividualResource loanPolicy = loanPoliciesFixture.canCirculateRolling();
+    final IndividualResource overdueFinePolicy = overdueFinePoliciesFixture.facultyStandard();
+    final IndividualResource lostItemFeePolicy = lostItemFeePoliciesFixture.facultyStandard();
+
+    ArrayList<RequestType> allowedRequestPolicyTypes = new ArrayList<>();
+    Arrays.stream(allowedRequestTypes.split(",")).forEach(type -> allowedRequestPolicyTypes.add(RequestType.from(type)));
```
[the remark this answers is not in the document]
Comment (bisecomsergiy) [reply src/test/java/api/requests/InstanceRequestsAPICreationTests.java]: Updated accordingly, thanks!

**Review by roman-barannyk — APPROVED** [review]

**Pull request merged**